packages/remix-server-runtime/responses.ts

```diff
@@ -1,3 +1,105 @@
+import { serializeError } from "./errors";
+
+export const DEFERRED_PROMISE_VALUE = "$$__REMIX_DEFERRED_PROMISE__$$";
+export const DEFERRED_CHUNK_SEPARATOR = "$$__REMIX_DEFERRED_PROMISE__$$";
+
+export type DeferredResponse = Response & {
+  deferred: Record<string, Promise<unknown>>;
+  initialData: unknown;
+};
+
+export function isDeferredResponse(value: any): value is DeferredResponse {
+  return (
+    isResponse(value) &&
+    typeof (value as DeferredResponse).deferred !== "undefined"
+  );
+}
+
+export type DeferredFunction = <Data>(
+  data: Data,
+  init?: number | ResponseInit
+) => Response;
+
+export const deferred: DeferredFunction = (data, init = {}) => {
+  class DeferredResponse extends Response {
+    public deferred: Record<string, Promise<unknown>>;
+    public initialData: unknown;
+
+    constructor(data: unknown, init?: ResponseInit) {
+      let deferred: Record<string, Promise<unknown>> = {};
+      let initialData = data;
+      if (typeof data === "object" && data !== null) {
+        let dataWithoutPromises = {} as Record<string, unknown>;
+
+        for (let [key, value] of Object.entries(data)) {
+          if (typeof value?.then === "function") {
+            deferred[key] = value;
+            dataWithoutPromises[key] = DEFERRED_PROMISE_VALUE + key;
+          } else {
+            dataWithoutPromises[key] = value;
+          }
+        }
+
+        initialData = dataWithoutPromises;
+      }
+
+      let encoder = new TextEncoder();
```

Comment:
If we can have some util method for doing the encoding I think that would help with code readability.

```ts
// Give me the stuff I need to create the Response
let { initialData, deferredData, body } = encodeRemixDeferred(data);
```